Adding a value at the maximum allowed index causes an OOM

Consider the following program:
```ballerina
public function main(string... args) {
    int[] a = [];
    a[2147483647-9] = 10; 
}
```
The above program causes an out of memory error. 
```
[2019-01-12 01:05:10,301] ERROR {org.ballerinalang.bre.bvm.BVMScheduler} - Java heap space 
java.lang.OutOfMemoryError: Java heap space
	at java.util.Arrays.copyOf(Arrays.java:3308)
	at org.ballerinalang.model.values.BValueArray.grow(BValueArray.java:513)
	at org.ballerinalang.model.values.BNewArray.ensureCapacity(BNewArray.java:122)
	at org.ballerinalang.model.values.BNewArray.prepareForAdd(BNewArray.java:82)
	at org.ballerinalang.model.values.BValueArray.add(BValueArray.java:210)
	at org.ballerinalang.bre.bvm.BVM.execStoreOpcodes(BVM.java:1769)
	at org.ballerinalang.bre.bvm.BVM.execute(BVM.java:330)
	at org.ballerinalang.bre.bvm.BVMScheduler.execute(BVMScheduler.java:72)
	at org.ballerinalang.bre.bvm.BVMExecutor.execute(BVMExecutor.java:184)
	at org.ballerinalang.bre.bvm.BVMExecutor.executeEntryFunction(BVMExecutor.java:92)
	at org.ballerinalang.BLangProgramRunner.runEntryFunc(BLangProgramRunner.java:74)
	at org.ballerinalang.launcher.LauncherUtils.runMain(LauncherUtils.java:171)
	at org.ballerinalang.launcher.LauncherUtils.runProgram(LauncherUtils.java:163)
	at org.ballerinalang.launcher.Main$RunCmd.execute(Main.java:302)
	at org.ballerinalang.launcher.Main$$Lambda$8/1510067370.accept(Unknown Source)
	at java.util.Optional.ifPresent(Optional.java:159)
	at org.ballerinalang.launcher.Main.main(Main.java:67)
 
[2019-01-12 01:05:10,317] ERROR {org.ballerinalang.launcher.Main} - java.lang.OutOfMemoryError: Java heap space 
java.lang.RuntimeException: java.lang.OutOfMemoryError: Java heap space
	at org.ballerinalang.bre.bvm.BVMScheduler.execute(BVMScheduler.java:77)
	at org.ballerinalang.bre.bvm.BVMExecutor.execute(BVMExecutor.java:184)
	at org.ballerinalang.bre.bvm.BVMExecutor.executeEntryFunction(BVMExecutor.java:92)
	at org.ballerinalang.BLangProgramRunner.runEntryFunc(BLangProgramRunner.java:74)
	at org.ballerinalang.launcher.LauncherUtils.runMain(LauncherUtils.java:171)
	at org.ballerinalang.launcher.LauncherUtils.runProgram(LauncherUtils.java:163)
	at org.ballerinalang.launcher.Main$RunCmd.execute(Main.java:302)
	at java.util.Optional.ifPresent(Optional.java:159)
	at org.ballerinalang.launcher.Main.main(Main.java:67)
Caused by: java.lang.OutOfMemoryError: Java heap space
	at java.util.Arrays.copyOf(Arrays.java:3308)
	at org.ballerinalang.model.values.BValueArray.grow(BValueArray.java:513)
	at org.ballerinalang.model.values.BNewArray.ensureCapacity(BNewArray.java:122)
	at org.ballerinalang.model.values.BNewArray.prepareForAdd(BNewArray.java:82)
	at org.ballerinalang.model.values.BValueArray.add(BValueArray.java:210)
	at org.ballerinalang.bre.bvm.BVM.execStoreOpcodes(BVM.java:1769)
	at org.ballerinalang.bre.bvm.BVM.execute(BVM.java:330)
	at org.ballerinalang.bre.bvm.BVMScheduler.execute(BVMScheduler.java:72)
	at org.ballerinalang.bre.bvm.BVMExecutor.execute(BVMExecutor.java:184)
	at org.ballerinalang.bre.bvm.BVMExecutor.executeEntryFunction(BVMExecutor.java:92)
	at org.ballerinalang.BLangProgramRunner.runEntryFunc(BLangProgramRunner.java:74)
	at org.ballerinalang.launcher.LauncherUtils.runMain(LauncherUtils.java:171)
	at org.ballerinalang.launcher.LauncherUtils.runProgram(LauncherUtils.java:163)
	at org.ballerinalang.launcher.Main$RunCmd.execute(Main.java:302)
	at org.ballerinalang.launcher.Main$$Lambda$8/1510067370.accept(Unknown Source)
	... 2 more
```

--- a/bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/scheduling/Scheduler.java
+++ b/bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/scheduling/Scheduler.java
@@ -22,6 +22,7 @@ import io.ballerina.runtime.api.async.StrandMetadata;
 import io.ballerina.runtime.api.constants.RuntimeConstants;
 import io.ballerina.runtime.api.creators.ErrorCreator;
 import io.ballerina.runtime.api.types.Type;
+import io.ballerina.runtime.api.utils.StringUtils;
 import io.ballerina.runtime.api.values.BError;
 import io.ballerina.runtime.api.values.BFunctionPointer;
 import io.ballerina.runtime.api.values.BObject;
@@ -438,6 +439,11 @@ public class Scheduler {
     private Throwable createError(Throwable t) {
         if (t instanceof StackOverflowError) {
             BError error = ErrorCreator.createError(BallerinaErrorReasons.STACK_OVERFLOW_ERROR);
+            error.setStackTrace(t.getStackTrace());
+            return error;
+        } else if (t instanceof OutOfMemoryError) {
+            BError error = ErrorCreator.createError(BallerinaErrorReasons.JAVA_OUT_OF_MEMORY_ERROR,
+                    StringUtils.fromString(t.getMessage()));
             error.setStackTrace(t.getStackTrace());
             return error;
         }

--- a/bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/util/exceptions/BallerinaErrorReasons.java
+++ b/bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/util/exceptions/BallerinaErrorReasons.java
@@ -54,6 +54,8 @@ public class BallerinaErrorReasons {
     public static final BString JAVA_NULL_REFERENCE_ERROR =
             StringUtils.fromString(BALLERINA_PREFIX.concat("JavaNullReferenceError"));
     public static final String JAVA_CLASS_NOT_FOUND_ERROR = BALLERINA_PREFIX.concat("JavaClassNotFoundError");
+    public static final BString JAVA_OUT_OF_MEMORY_ERROR =
+            StringUtils.fromString(BALLERINA_PREFIX.concat("OutOfMemoryError"));
 
     public static final BString BALLERINA_PREFIXED_CONVERSION_ERROR =
             StringUtils.fromString(BALLERINA_PREFIX.concat("ConversionError"));

--- a/tests/ballerina-test-utils/src/main/java/org/ballerinalang/test/BRunUtil.java
+++ b/tests/ballerina-test-utils/src/main/java/org/ballerinalang/test/BRunUtil.java
@@ -1316,7 +1316,11 @@ public class BRunUtil {
     }
 
     public static String runMain(CompileResult compileResult, String... args) {
-        ExitDetails exitDetails = run(compileResult, args);
+        return runMain(compileResult, new ArrayList<>(), args);
+    }
+
+    public static String runMain(CompileResult compileResult, List<String> javaOpts, String... args) {
+        ExitDetails exitDetails = run(compileResult, javaOpts, args);
         if (exitDetails.exitCode != 0) {
             throw new RuntimeException(exitDetails.errorOutput);
         }
@@ -1324,6 +1328,10 @@ public class BRunUtil {
     }
 
     public static ExitDetails run(CompileResult compileResult, String... args) {
+        return run(compileResult, new ArrayList<>(), args);
+    }
+
+    public static ExitDetails run(CompileResult compileResult, List<String> javaOpts, String... args) {
         PackageManifest packageManifest = compileResult.packageManifest();
         String initClassName = JarResolver.getQualifiedClassName(packageManifest.org().toString(),
                 packageManifest.name().toString(),
@@ -1333,12 +1341,15 @@ public class BRunUtil {
 
         try {
             final List<String> actualArgs = new ArrayList<>();
-            actualArgs.add(0, "java");
-            actualArgs.add(1, "-Xmx2048m");
-            actualArgs.add(2, "-cp");
+            int index = 0;
+            actualArgs.add(index++, "java");
+            for (int i = 0; i < javaOpts.size(); i++) {
+                actualArgs.add(index++, javaOpts.get(i));
+            }
+            actualArgs.add(index++, "-cp");
             String classPath = System.getProperty("java.class.path") + ":" + getClassPath(classLoader);
-            actualArgs.add(3, classPath);
-            actualArgs.add(4, initClassName);
+            actualArgs.add(index++, classPath);
+            actualArgs.add(index++, initClassName);
             actualArgs.addAll(Arrays.asList(args));
 
             final Runtime runtime = Runtime.getRuntime();

--- a/tests/ballerina-test-utils/src/main/java/org/ballerinalang/test/BRunUtil.java
+++ b/tests/ballerina-test-utils/src/main/java/org/ballerinalang/test/BRunUtil.java
@@ -1334,10 +1334,11 @@ public class BRunUtil {
         try {
             final List<String> actualArgs = new ArrayList<>();
             actualArgs.add(0, "java");
-            actualArgs.add(1, "-cp");
+            actualArgs.add(1, "-Xmx2048m");
+            actualArgs.add(2, "-cp");
             String classPath = System.getProperty("java.class.path") + ":" + getClassPath(classLoader);
-            actualArgs.add(2, classPath);
-            actualArgs.add(3, initClassName);
+            actualArgs.add(3, classPath);
+            actualArgs.add(4, initClassName);
             actualArgs.addAll(Arrays.asList(args));
 
             final Runtime runtime = Runtime.getRuntime();

--- a/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/jvm/ErrorTest.java
+++ b/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/jvm/ErrorTest.java
@@ -22,7 +22,6 @@ import org.ballerinalang.core.model.values.BValue;
 import org.ballerinalang.test.BCompileUtil;
 import org.ballerinalang.test.BRunUtil;
 import org.ballerinalang.test.CompileResult;
-import org.testng.Assert;
 import org.testng.annotations.AfterClass;
 import org.testng.annotations.BeforeClass;
 import org.testng.annotations.Test;
@@ -76,15 +75,12 @@ public class ErrorTest {
         BRunUtil.invoke(compileResult, "testSelfReferencingError");
     }
 
-    @Test
+    @Test(description = "Test runtime out of memory error", expectedExceptions = java.lang.RuntimeException.class,
+            expectedExceptionsMessageRegExp = "error: \\{ballerina}OutOfMemoryError \\{\"message\":\"Java heap " +
+                    "space\"}.*")
     public void testRuntimeOOMError() {
-        try {
-            CompileResult compileResult = BCompileUtil.compile("test-src/jvm/runtime-oom-error.bal");
-            BRunUtil.runMain(compileResult, new String[]{});
-        } catch (Throwable e) {
-            return;
-        }
-        Assert.fail("runtime out of memory errors are not handled");
+        CompileResult compileResult = BCompileUtil.compile("test-src/jvm/runtime-oom-error.bal");
+        BRunUtil.runMain(compileResult);
     }
 
     @AfterClass

--- a/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/jvm/ErrorTest.java
+++ b/tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/jvm/ErrorTest.java
@@ -26,6 +26,9 @@ import org.testng.annotations.AfterClass;
 import org.testng.annotations.BeforeClass;
 import org.testng.annotations.Test;
 
+import java.util.ArrayList;
+import java.util.List;
+
 /**
  * Test cases to cover error related tests on JBallerina.
  *
@@ -80,7 +83,10 @@ public class ErrorTest {
                     "space\"}.*")
     public void testRuntimeOOMError() {
         CompileResult compileResult = BCompileUtil.compile("test-src/jvm/runtime-oom-error.bal");
-        BRunUtil.runMain(compileResult);
+        final List<String> javaOpts = new ArrayList<>();
+        javaOpts.add(0, "-Xms256m");
+        javaOpts.add(1, "-Xmx256m");
+        BRunUtil.runMain(compileResult, javaOpts);
     }
 
     @AfterClass

